### ISSUE TITLE
Targeting an arch in gcc but not in LLVM.

### ISSUE THREAD

**Issue by cr1901** [antoyo/rustc_codegen_gcc#49]
I'm on vacation till Wednesday, but I figured I should ask this question now while it's fresh on my mind:

I'm interested in getting Rust support as soon as possible for a few architectures with the new `gcc` codegen; SuperH would be the highest priority for me. What current obstacles exist for me to try out cross compiling to bare-metal `sh2`/`j2`, and reporting back (and maybe even fixing, if I can get a grasp on RTL!) any issues I can find. What obstacles exist for attempting to test `sh2` codegen using `rustc`? I can think of a few, but would love to know more:

1. I understand that distributing `libgccjit.so` is more complicated than the LLVM backend because each `libgccjit.so` can only target one arch. Has there been movement on how all the different `libgccjit.so`s should coexist?
   1. I don't remember where, but I heard `libgccjit.so` is difficult to use in a cross-compile setting; how true is that?
2. I'm not sure if `libcore` is tested for _any_ arch where `libstd` doesn't exist in Rust proper, let alone for `gcc` codegen backends. Is there a way to run the `libcore` tests in `qemu` or the `gdb` simulators? Testing `libcore` would be a good litmus test for bare-metal `sh2`/`j2`<sup>1</sup>.
3. Upstream `rustc` would have to have a target registered for `sh2`/`j2`, and would somehow need to know that only the `gcc` codegen backend is supported for now (LLVM backend might come later, but it's not a high priority). Does this "choose a backend" logic already exist in upstream Rust? Is there additional ABI work that needs to be done to add `sh2` upstream?

I assume 2. and 3. need to be handled first before I can start actual codegen for "archs which only have a gcc port". I'd be interested in getting the ball rolling there (at the very least testing, or contributing code depending on how much time I can allocate to this).

## Footnotes
1.  Right now the common Linux ABI seems to be MMU-less [`fdpic`](https://gcc.gnu.org/legacy-ml/gcc/2008-02/msg00619.html), and I have no idea whether `rustc` meaningfully supports anything like it.

**Comment by antoyo** [comment]
I'm on vacation too.

1. will be under discussion to see what we can do with rustup
2. I guess we could cross-compile the libcore tests and run them in qemu.
3. Not sure I understand. For the first part, you mean in rustup? For the "choose a backend", there's already a command line argument to select the gcc/cranelift codegen, is this what you mean? For the ABI, there might be some work needed in Rust indeed. There will also be work needed for the new inline assembly syntax.

**Comment by glaubitz** [comment]
I'm one of Debian's porters for the less common architectures such as m68k and sh4 (SuperH).

I can bootstrap `rustc` for any architecture in Debian once `rustc_codegen_gcc` has been upstreamed into `rustc` and the Rust compiler can be switched to use `libgccjit` rather than `libllvm`.

I guess we will have to add the various boilerplates for the additional architecture to `rustc` then once `rustc_codegen_gcc` has been upstreamed.

**Comment by cr1901** [comment]
I'm back from vacation.

>For the first part, you mean in rustup?

Not necessarily, `rustc` proper also supports the `--target` flag.

>For the "choose a backend", there's already a command line argument to select the gcc/cranelift codegen, is this what you mean?

What I mean is, something like `rustc --target=sh2` by itself should "just work" by defaulting to a `libgccjit` codegen at a suitable path, and error if the appropriate `libgccjit` is not installed. Additionally, something like `rustc --target=sh2 -Zcodegen-backend=/path/to/libllvm` should also error out because there is no `libllvm` that supports `sh2`. Though maybe logic like that already exists in `libllvm`?

@antoyo I guess my overarching question is: What can I do to start adding cross-compilation support to `rustc_codegen_gcc`, taking the above into account, without getting in your way?

**Comment by antoyo** [comment]
Sorry for the delay.

As far as I know `--target` is merely the target triple used and it has no logic to choose a backend. If you would like to change this behaviour (or add a new flag doing what you would like) I suggest you start a discussion somewhere (maybe on Zulip) and ultimately open a MCP to propose the change to the compiler team.

To start helping on the rustc side, you could start a discussion to see whether you can start adding the new architectures you would want in [this directory](https://github.com/rust-lang/rust/tree/master/compiler/rustc_target/src/spec) and the new inline-asm register support for those new architectures so that [this code](https://github.com/antoyo/rustc_codegen_gcc/blob/master/src/asm.rs#L568) can be completed with those new architectures.
Maybe you'll have to wait until the gcc codegen is merged, but this discussion can start now ;) .

**Comment by bjorn3** [comment]
The target spec used to have a field to specify which backend to use. This was introduced to allow emscripten to use a rustc_codegen_llvm linked against emscripten's llvm instead of the newer default llvm used by rustc. I removed this field in https://github.com/rust-lang/rust/pull/74394 as the emscripten support no longer needed a different llvm and thus the field wasn't used anymore.

**Comment by cr1901** [comment]
@bjorn3 Are you aware of any difficulties adding backend selection back, in similar to how it was (before it was removed)?

>Maybe you'll have to wait until the gcc codegen is merged, but this discussion can start now ;) .

@antoyo Things have changed on my end in the past 4 weeks, so right now I don't have the time I thought I would to commit to starting a backend port (besides testing x86). But I can at least open issues, get feedback, and study your code in the interim :). I just need a few weeks to get things sorted.

**Comment by antoyo** [comment]
Some notes from my attempts at running Rust code in a Debian m68k qemu virtual machine:

 * WIP PR adding notes for cross-compilation: #68 
 * The `object` crate will need to support the new [architectures](https://github.com/gimli-rs/object/blob/d26bc89508852bfe446e06ba5a22f0066ab9e1f1/src/common.rs#L5) in order to compile dylib.
 * The `libc` crate will need to support the new [architectures](https://github.com/rust-lang/libc/tree/master/src/unix/linux_like/linux/arch) (it seems to be my last blocker in order to run the tests from libcore on m68k).

I was able to run a simple program using core and a "hello world" using the std on m68k.

**Comment by cr1901** [comment]
@antoyo Excellent :D! In the coming days, I'll see if I can duplicate this setup in the coming days in [rcg_docker](https://github.com/cr1901/rcg_docker), which has bitrotted a bit since your rebase.

**Comment by MabezDev** [comment]
> The object crate will need to support the new [architectures](https://github.com/gimli-rs/object/blob/d26bc89508852bfe446e06ba5a22f0066ab9e1f1/src/common.rs#L5) in order to compile dylib.
The libc crate will need to support the new [architectures](https://github.com/rust-lang/libc/tree/master/src/unix/linux_like/linux/arch) (it seems to be my last blocker in order to run the tests from libcore on m68k).

Are these required for no_std builds? I'd like to take a stab at compiling for Xtensa using the GCC backend. Is it possible to statically compile `core` etc with just specifying `TARGET_TRIPLE`?

**Comment by bjorn3** [comment]
> Are these required for no_std builds?

The object crate needs to support it the new target. While in the past it was only necessary for dylibs, rustc now also uses object for writing object files containing the crate metadata in rlibs.

> Is it possible to statically compile core etc with just specifying TARGET_TRIPLE?

Once you add support to the object crate (and update rustc with the new version) and compile libgccjit to target xtensa (every compiled version of gcc can only target a single triple unlike llvm) it may work.

**Comment by antoyo** [comment]
@MabezDev Also, if you want to use inline assembly, that will require an update in rustc to support this architecture.

**Comment by glaubitz** [comment]
@antoyo Do you have a quick guide on how to bootstrap rustc using the backend for a given architecture?

I would love to give it a try with m68k which currently uses the work-in-progress LLVM backend.

**Comment by MabezDev** [comment]
> @MabezDev Also, if you want to use inline assembly, that will require an update in rustc to support this architecture.

Thanks for the hint! Fortunately I've already added Xtensa assembly support in our fork that uses a custom llvm backend: https://github.com/esp-rs/rust/commit/9d4ab85fba15755c4cfa14437f422421cd073d34

**Comment by antoyo** [comment]
> @antoyo Do you have a quick guide on how to bootstrap rustc using the backend for a given architecture?
> 
> I would love to give it a try with m68k which currently uses the work-in-progress LLVM backend.

That might be simpler to wait until upstream rustc is updated with the latest changes in this repository, but if you want to do it manually locally, here's how to do it.

Add the missing changes from rustc_codegen_gcc in your local rustc repo (might require to use the commit for nightly-2022-03-26).

Update the `config.toml` in the rustc repo to add the following config:
```toml
changelog-seen = 2

[rust]
codegen-backends = ["gcc"]
deny-warnings = false
new-symbol-mangling = true
```
then, run the following command:
```
LIBRARY_PATH=/path/to/libgccjit_dir LD_LIBRARY_PATH=/path/to/libgccjit_dir ./x.py install -j4 --stage 1 --host m68k-unknown-linux-gnu --target m68k-unknown-linux-gnu
```
I haven't tried cross-compiling rustc yet, so that might not work.

**Comment by bjorn3** [comment]
> I haven't tried cross-compiling rustc yet, so that might not work.

It probably won't work. Rustc needs to be compiled for the build machine in addition to the chosen host to build the standard library. This would require some way to choose the right libgccjit for depending on the triple to compile for.

**Comment by glaubitz** [comment]
> Rustc needs to be compiled for the build machine in addition to the chosen host to build the standard library.

Wouldn't it be possible to build a Rust cross-compiler with the standard Rust compiler on the host system?

> This would require some way to choose the right libgccjit for depending on the triple to compile for.

On Debian unstable, we should have libgccjit versions for all targets.

**Comment by bjorn3** [comment]
> Wouldn't it be possible to build a Rust cross-compiler with the standard Rust compiler on the host system?

The standard library needs to be compiled by the exact same rustc version as the one that actually uses it. This means that we first have to build stage1 rustc for the host using the bootstrap compiler. This stage1 rustc can then build libstd. And finally another stage1 rustc can be built for the target which can then load the aformentioned libstd version.

**Comment by antoyo** [comment]
> > Wouldn't it be possible to build a Rust cross-compiler with the standard Rust compiler on the host system?
> 
> The standard library needs to be compiled by the exact same rustc version as the one that actually uses it. This means that we first have to build stage1 rustc for the host using the bootstrap compiler. This stage1 rustc can then build libstd. And finally another stage1 rustc can be built for the target which can then load the aformentioned libstd version.

Do you have any suggestion on how we could do that, perhaps by specifying multiple libgccjit.so paths in the config file?

**Comment by bjorn3** [comment]
That would be an option.

**Comment by glaubitz** [comment]
One option might be to build the suitable libgccjit version during the build process of `rustc` itself similar to what's already done with LLVM.

Would that be feasible?

**Comment by antoyo** [comment]
@glaubitz Continuing the discussion from #123 here:

> m68k support in `libc` is complete now, but the `rustc` source tree has not updated to the latest version of `libc` due to a Illumos-related change breaking the CI (it's being worked on).

One thing you might try is editing the file `build_sysroot/sysroot_src/library/std/Cargo.toml` (that you get after running `prepare.sh`) in the repo of rustc_codegen_gcc to use the latest version of `libc`:
```toml
[patch.crates-io]
libc = { git = "https://github.com/rust-lang/libc", default-features = false, features = ['rustc-dep-of-std'] }
```

That *might* work.

**Comment by glaubitz** [comment]
> @glaubitz Continuing the discussion from #123 here:
> 
> That _might_ work.

I think I can just update the `libc` crate in my local branch using `cargo update -p libc --precise`, so that's probably not the biggest issue.

The only thing that I haven't understood yet: From where do I get a libgccjit that can cross-compile for a given architecture? Do I build libgccjit from source for that purpose?

**Comment by antoyo** [comment]
@glaubitz One more thing, in case you don't know already: use my [fork of gcc](https://github.com/antoyo/gcc) as some of my patches were not merged yet upstream.

**Comment by dkm** [comment]
FYI, I've merged the support for libgccjit in crosstool-ng (https://github.com/crosstool-ng/crosstool-ng/pull/1873). This may help in getting the lib for some other target (initial goal was to have a cross rustc_cg_gcc in compiler-explorer https://github.com/compiler-explorer/compiler-explorer/issues/3760)

**Comment by leet-hakker** [comment]
Is there any update on this? I'm interested in the SuperH architecture exactly as well, does the opener or anyone else have any information/progress?

**Comment by glaubitz** [comment]
> Is there any update on this? I'm interested in the SuperH architecture exactly as well, does the opener or anyone else have any information/progress?

If you're interested in SuperH, you might want to join #linux-sh on Libera IRC as well as #debian-ports on OFTC IRC.

**Comment by antoyo** [comment]
> Is there any update on this? I'm interested in the SuperH architecture exactly as well, does the opener or anyone else have any information/progress?

I put in the [wiki](https://github.com/rust-lang/rustc_codegen_gcc/wiki#superh) some information about someone who was able to compile some Rust code for SuperH.

**Comment by glaubitz** [comment]
> > Is there any update on this? I'm interested in the SuperH architecture exactly as well, does the opener or anyone else have any information/progress?
> 
> I put in the [wiki](https://github.com/rust-lang/rustc_codegen_gcc/wiki#superh) some information about someone who was able to compile some Rust code for SuperH.

Wow, very cool. I guess we can start adding non-LLVM architectures to Rust's libc and std libraries.

**Comment by leet-hakker** [comment]
> > Is there any update on this? I'm interested in the SuperH architecture exactly as well, does the opener or anyone else have any information/progress?
> 
> I put in the [wiki](https://github.com/rust-lang/rustc_codegen_gcc/wiki#superh) some information about someone who was able to compile some Rust code for SuperH.

That's really interesting! It turns out that the author of that Dockerfile has embarked on exactly the same project I have. I'll have a look at the project. Thanks.

**Comment by antoyo** [comment]
@glaubitz wrote:

> @antoyo Do you have a guide on how to get started adding a new architecture to rustc using rustc_codegen_gcc?
>
> I have heard from people in the Dreamcast community that they had success with sh-elf using rustc_codegen_gcc, so I would like to add sh4-unknown-linux-gnu using this backend.

Let's continue the discussion here instead of in #123 since that issue is closed (or you could open a new issue if you prefer).

I have no idea how to do this, but I can work with you to find out this information and to make the changes to rustc if necessary. You could start by asking on Zulip in the appropriate stream and cc me or I can do that if you prefer.

**Comment by antoyo** [comment]
@glaubitz: I have a working cross-compiler targetting `m68k-unknown-linux-gnu` from x86-64.
You mean the patches to GCC? Not, I still have a bunch to merge for gcc 14 (and the one necessary for unwinding will not be merged for gcc 14).

I could try building a native `rustc` for `m68k-unknown-linux-gnu`, but it is very likely it won't be able to compile many programs (I haven't tried on x86 for a long time, but back then, that was an issue). Would that still be useful for you?

 Did you try to cross-compile other softwares (like `librsvg` or `cargo`) for m68k?

**Comment by glaubitz** [comment]
> Some notes from my attempts at running Rust code in a Debian m68k qemu virtual machine:
> 
>     * WIP PR adding notes for cross-compilation: [Add notes for cross-compilation to gcc-only targets #68](https://github.com/rust-lang/rustc_codegen_gcc/pull/68)
> 
>     * The `object` crate will need to support the new [architectures](https://github.com/gimli-rs/object/blob/d26bc89508852bfe446e06ba5a22f0066ab9e1f1/src/common.rs#L5) in order to compile dylib.
> 
>     * The `libc` crate will need to support the new [architectures](https://github.com/rust-lang/libc/tree/master/src/unix/linux_like/linux/arch) (it seems to be my last blocker in order to run the tests from libcore on m68k).
> 
> 
> I was able to run a simple program using core and a "hello world" using the std on m68k.

m68k support has been added to `libc` in https://github.com/rust-lang/libc/commit/7a027d34c2d748b23a81fb75c9b84242fe70fcda. Support in `object` is missing, but I can take care of that, too.

What I still don't understand: How do we make `rustc` aware of an architecture which is not supported by `rustc_codegenllvm`? This would be the first step before we can add support for these new architectures to `libc` as the `libc` crate can only add targets that `rustc` recognizes.

**Comment by GuillaumeGomez** [comment]
Maybe by adding a vec into the target info: `supported_by: Vec<Backend>` where `Backend` would be an enum composed of `Llvm`, `Gcc` and `Cranelift`?

**Comment by glaubitz** [comment]
> Maybe by adding a vec into the target info: `supported_by: Vec<Backend>` where `Backend` would be an enum composed of `Llvm`, `Gcc` and `Cranelift`?

Sounds plausible to me. Do you have a rough idea how this would look like?

**Comment by GuillaumeGomez** [comment]
A new field in `base` that is used for all targets (example [here](https://github.com/rust-lang/rust/blob/master/compiler/rustc_target/src/spec/targets/aarch64_apple_darwin.rs)) I suppose?

**Comment by dkm** [comment]
Have you checked https://rustc-dev-guide.rust-lang.org/building/new-target.html ?

**Comment by glaubitz** [comment]
> Have you checked https://rustc-dev-guide.rust-lang.org/building/new-target.html ?

I have added `rustc_codegen_llvm` targets before. But I have no clue how to wire up `rustc_codegen_gcc` targets.

And, more surprisingly, the [instructions](https://dreamcast.wiki/Rust_on_Dreamcast) to build `rustc` for the Dreamcast using the `sh-elf` don't even seem to patch the Rust tree in any way to add support for this target.

I would expect that we would need to add spec files to `rustc_target` at least.

**Comment by darcagn** [comment]
> > Have you checked https://rustc-dev-guide.rust-lang.org/building/new-target.html ?
> 
> I have added `rustc_codegen_llvm` targets before. But I have no clue how to wire up `rustc_codegen_gcc` targets.
> 
> And, more surprisingly, the [instructions](https://dreamcast.wiki/Rust_on_Dreamcast) to build `rustc` for the Dreamcast using the `sh-elf` don't even seem to patch the Rust tree in any way to add support for this target.
> 
> I would expect that we would need to add spec files to `rustc_target` at least.

I'm using a little bit of a trick for that in my scripts that wrap the rustc_codegen_gcc tools. `rustc` thinks it's compiling for MIPS, and so it outputs object files that contain the SH code from GCC but with MIPS architecture specified in the header. The script then rewrites to SH before sending these as input to the linker. 

As for the `libc` crate, I've been able to add support for our homebrew KallistiOS on Dreamcast because I am gating the configuration behind a `kallistios` OS target in the target JSON file and in forked `libc`/`rust` crates and not worrying about architecture-specific configurations. I haven't published these forks yet but I will do so in the coming weeks as time permits.

It has been working fine doing this so far, but obviously this isn't the right way to handle things and we'll need to add SH, etc. to `rustc` properly in the long run.

**Comment by glaubitz** [comment]
> I'm using a little bit of a trick for that in my scripts that wrap the rustc_codegen_gcc tools. `rustc` thinks it's compiling for MIPS, and so it outputs object files that contain the SH code from GCC but with MIPS architecture specified in the header. The script then rewrites to SH before sending these as input to the linker.

OK, that clarifies my confusion. Thanks a lot!

> As for the `libc` crate, I've been able to add support for our homebrew KallistiOS on Dreamcast because I am gating the configuration behind a `kallistios` OS target in the target JSON file and in forked `libc`/`rust` crates and not worrying about architecture-specific configurations. I haven't published these forks yet but I will do so in the coming weeks as time permits.

Adding `libc` is relatively easy if you're just adding a new Linux target (which is what I'm interested in).

> It has been working fine doing this so far, but obviously this isn't the right way to handle things and we'll need to add SH, etc. to `rustc` properly in the long run.

Would be great to document this as well as possible to ease the process of adding new `libgccjit` targets for others.

**Comment by gyrovorbis** [comment]
I'm not super familiar with what exactly @darcagn had to go through to get all of this set up with rustc_codegen_gcc targeting SH4, but I have been on the KallistiOS side helping to support the back-end needs for the Rust stdlib and even tokio async library on the Sega Dreamcast, working on providing a pthread layer of abstraction above our own custom kernel among a few other things in C...

And holy crap, guys, I can't believe how well this stuff works and just how quickly @darcagn was able to work his way up to having async and advanced Rust concurrency running just fine on the Sega Dreamcast with our TLS model (currently only supports local-exec), our atomics (we provide our own just toggling interrupts), and many other things...

I really think this is going to be able to bring a lot of epic Rust stuff to the SH arch, and I would love to see us getting support for our target upstreamed.... 

Also its's so nice to see all of you SH rockstars here and how much interest there is in bringing this cool new stuff to the target. I'll have to join the IRC server!

**Comment by glaubitz** [comment]
> I'm not super familiar with what exactly @darcagn had to go through to get all of this set up with rustc_codegen_gcc targeting SH4, but I have been on the KallistiOS side helping to support the back-end needs for the Rust stdlib and even tokio async library on the Sega Dreamcast, working on providing a pthread layer of abstraction above our own custom kernel among a few other things in C...
> 
> And holy crap, guys, I can't believe how well this stuff works and just how quickly @darcagn was able to work his way up to having async and advanced Rust concurrency running just fine on the Sega Dreamcast with our TLS model (currently only supports local-exec), our atomics (we provide our own just toggling interrupts), and many other things...

This gives me hope that we can get `rustc_codegen_gcc` get to work for alpha, hppa, m68k and sh4 in Debian as well.

> I really think this is going to be able to bring a lot of epic Rust stuff to the SH arch, and I would love to see us getting support for our target upstreamed....
> 
> Also its's so nice to see all of you SH rockstars here and how much interest there is in bringing this cool new stuff to the target. I'll have to join the IRC server!

Please keep in mind that there are still a lot of open issues with the SH backend in GCC, so please help me [organize a crowdfunding campaign](https://dcemulation.org/phpBB/viewtopic.php?t=106838) to pay an experienced GCC developer to revamp the SH backend in GCC. The SH backend really isn't in a good shape at the moment with lots of source packages triggering ICEs.

**Comment by glaubitz** [comment]
@antoyo Are there plans to add GCC upstream as a git submodule to the rustc git tree similar to LLVM?

I think it would be great to have `rustc_codegen_gcc` eventually work exactly like `rustc_codegen_llvm`, so that in the end the build system will automatically build `libgccjit` for targets that have GCC backends only.

**Comment by GuillaumeGomez** [comment]
https://github.com/rust-lang/rust/pull/125419

**Comment by glaubitz** [comment]
> [rust-lang/rust#125419](https://github.com/rust-lang/rust/pull/125419)

Woohoo! Awesome, thanks!

**Comment by dkm** [comment]
Looks like there's some confusion about the GPL license in the comments. Hope this won't push it back or trigger false alarm :). Nice to see it included in rustc :thumbup:

**Comment by GuillaumeGomez** [comment]
Don't hesitate to comment to clear confusion. ;)

**Comment by antoyo** [comment]
> @antoyo Are there plans to add GCC upstream as a git submodule to the rustc git tree similar to LLVM?
> 
> I think it would be great to have `rustc_codegen_gcc` eventually work exactly like `rustc_codegen_llvm`, so that in the end the build system will automatically build `libgccjit` for targets that have GCC backends only.

I'm not sure how you want to install rustc_codegen_gcc, but the above might be to get it via rustup: would that be alright?
@GuillaumeGomez can comment to confirm.

**Comment by GuillaumeGomez** [comment]
It's the plan.

**Comment by glaubitz** [comment]
> > @antoyo Are there plans to add GCC upstream as a git submodule to the rustc git tree similar to LLVM?
> > I think it would be great to have `rustc_codegen_gcc` eventually work exactly like `rustc_codegen_llvm`, so that in the end the build system will automatically build `libgccjit` for targets that have GCC backends only.
> 
> I'm not sure how you want to install rustc_codegen_gcc, but the above might be to get it via rustup: would that be alright? @GuillaumeGomez can comment to confirm.

I'm not sure how that's related to `rustup`? I'm talking about being able to build `rustc_codegen_gcc` from git.

To be honest, I have never used `rustup` as either use distribution packages or build the compiler from git.

**Comment by GuillaumeGomez** [comment]
Nothing prevents you to build gcc from our fork.

**Comment by glaubitz** [comment]
> Nothing prevents you to build gcc from our fork.

A forked GCC won't help me when I want to build a stand-alone compiler for alpha, hppa, m68k and sh4.

My hope was that `rustc_codegen_gcc` would soon help us to bring Rust to more architectures.

**Comment by GuillaumeGomez** [comment]
I'm getting more and more confused by what you want to do here. GCC upstream doesn't have all the patches we wrote that are needed for rust, so for now only our fork can be used with `rustc_codegen_gcc`. An important note about cross-compilation with GCC (I don't know how much you know about it so just in case...): GCC only allows to compile to one target. You need to compile GCC for this target on your current platform for it to be used to build cross-compilation. It's an annoying limitation and has nothing to do with Rust, so again, you need to build your own GCC. Only difference: you need to build the one from our fork (which is the submodule being added to the rust repository).

**Comment by cr1901** [comment]
So lately, I've been just lurking, but I still read this issue:

>I understand that distributing libgccjit.so is more complicated than the LLVM backend because each libgccjit.so can only target one arch. Has there been movement on how all the different libgccjit.sos should coexist?

When I asked this, something that I intended to imply asking is "will upstream Rust handle building/providing libgccjits, so that a user need not compile their own"? I take it from "Nothing prevents you to build gcc from our fork." that if someone wants to use the gcc backend, they are expected to provide their own gcc, and if someone wants there to be a turnkey solution to installing the gcc backend, a third-party will have to host libgccit.so's (or create a project for providing them all)?

**Comment by GuillaumeGomez** [comment]
> When I asked this, something that I intended to imply asking is "will upstream Rust handle building/providing libgccjits, so that a user need not compile their own"?

We plan to provide libgccjit.so and gcc for the target you asked, but nothing for cross-compilation (and linux-only for now).

> I take it from "Nothing prevents you to build gcc from our fork." that if someone wants to use the gcc backend, they are expected to provide their own gcc, and upstream Rust has no plans to host the libgccjit shared objects? I.e. if someone wants there to be a turnkey solution to installing the gcc backend from `rustup`, a third-party will have to host libgccit.so's (or some other project will need to be created for providing them all)?

Like I mentioned above, `rustup` will provide needed files to use the Rust GCC backend. Without `rustup`, it's entirely up to the user indeed.

**Comment by glaubitz** [comment]
> ﻿I'm getting more and more confused by what you want to do here. GCC upstream doesn't have all the patches we wrote that are needed for rust, so for now only our fork can be used with rustc_codegen_gcc. An important note about cross-compilation with GCC (I don't know how much you know about it so just in case...): GCC only allows to compile to one target. You need to compile GCC for this target on your current platform for it to be used to build cross-compilation. It's an annoying limitation and has nothing to do with Rust, so again, you need to build your own GCC. Only difference: you need to build the one from our fork (which is the submodule being added to the rust repository).

Well, my point is that I don’t understand why the necessary patches haven’t been upstreamed yet.

I don’t need a cross-compiler for all targets, I just want to be able to build stand-alone rustc for new targets.

I had hoped this to be much easier as I remember this was stated as one of the big advantages over gccrs which I have already used to build minimal rust binaries for new targets.

**Comment by GuillaumeGomez** [comment]
Well, quite a lot of them are waiting for reviews. Getting things merged it taking veeeeeeery long.

**Comment by antoyo** [comment]
> Well, my point is that I don’t understand why the necessary patches haven’t been upstreamed yet. I don’t need a cross-compiler for all targets, I just want to be able to build stand-alone rustc for new targets. I had hoped this to be much easier as I remember this was stated as one of the big advantages over gccrs which I have already used to build minimal rust binaries for new targets.

The usual rustc also use a forked LLVM, if I'm not mistaken. Is there any reason why this wouldn't suit you that it is the same (for the time being: in the long run, the plan is to use upstream libgccjit, but the reviews take a long time to be done)?

Could you please explain how you would want to build rustc_codegen_gcc since you seem to want something different than most people (that want to download it via `rustup`)? Please describe the commands you would want to write to build it.

**Comment by bjorn3** [comment]
> The usual rustc also use a forked LLVM, if I'm not mistaken.

It also builds with upstream LLVM. The forked LLVM just has some more bugfixes, not any actually new non-optional features.

**Comment by cr1901** [comment]
>Like I mentioned above

Yea, it occurred to me that I asked my question prematurely and hoped you wouldn't notice :P.

>Like I mentioned above, rustup will provide needed files to use the Rust GCC backend. Without rustup, it's entirely up to the user indeed.

All of this is perfectly fine with me. I have scripts (actually a Rust crate using `rust_cmd_lib`) to build LLVM and rustc from git. I have plans to add building libgccjit from it as well.

My scripts assume a shared _external_ LLVM checkout that I use for other projects, and has logic to handle swapping between upstream and Rust branches. Plan would be to do the same for the forked gcc repo. @glaubitz if your concern is not having 10 million gcc and llvm checkouts, I've used my script successfully for ~2 years now, building Rust once a week, and should probably open-source it. It's like 300 lines of Rust if that, and is mainly Rust to handle the git checkout logic.

>We plan to provide libgccjit.so and gcc for the target you asked, but nothing for cross-compilation (and linux-only for now).

~~FWIW, in my initial post, I asked for bare-metal SuperH target. So we're getting SuperH support in `rustup` :P?~~

**Comment by glaubitz** [comment]
Has there been any progress on this particular issue? The pressure for Rust support is building up and it would be nice if we could start adding architectures to the Rust compiler which are exclusively supported by `libgccjit`.

**Comment by GuillaumeGomez** [comment]
Nothing prevents you to do so. You need to declare the architecture like done [in this PR](https://github.com/gimli-rs/object/pull/762/files).

**Comment by glaubitz** [comment]
> Nothing prevents you to do so. You need to declare the architecture like done [in this PR](https://github.com/gimli-rs/object/pull/762/files).

Well, in the past this didn't work. We were already on it with [hppa support in rust-libc](https://github.com/rust-lang/libc/pull/3539) and [it got pulled again](https://github.com/rust-lang/libc/pull/3542).

Hence my question was whether it works these days.

**Comment by GuillaumeGomez** [comment]
As always: communications. Talk with the compiler team and the libc team (on [rust's zulip](https://rust-lang.zulipchat.com)) to ensure they're ok with you adding that. Then things should be much smoother.

**Comment by glaubitz** [comment]
> As always: communications. Talk with the compiler team and the libc team (on [rust's zulip](https://rust-lang.zulipchat.com)) to ensure they're ok with you adding that. Then things should be much smoother.

Well, my question is whether it's technically already possible to add an architecture that is not supported by `rustc_codegen_llvm` but just by `rustc_codegen_gcc` (or `rustc_codegen_cranelift`)?

In the past, that wasn't the case which is why this issue was opened in the first place.

**Comment by GuillaumeGomez** [comment]
I'm not sure but I think the first link I shared allows it (to be confirmed).

**Comment by thesamesam** [comment]
Do you know who could confirm it, please? We've been through this a few times before, then we find the policy isn't what people think it is.

> As always: communications. Talk with the compiler team and the libc team (on [rust's zulip](https://rust-lang.zulipchat.com/)) to ensure they're ok with you adding that. Then things should be much smoother.

It's not like we merged the PR ourselves.

**Comment by GuillaumeGomez** [comment]
Then it's internal team communications. Not sure what's the best course here except asking the rust project leaders?

**Comment by thesamesam** [comment]
Can they chip in here please then? I don't know who to ping.

**Comment by antoyo** [comment]
I'd say the situation got worse recently since it is now required that the `object` crate supports the architecture for us to be able to compile to an architecture not supported by `rustc_codegen_llvm`.
But, I saw you just made a [PR for HPPA](https://github.com/gimli-rs/object/pull/793), so it's a first step in this direction.

Let me try to find the right people to communicate to and ask about this, but I need to know a few information about what exactly you want:

 * How do you intend to build `rustc_codegen_gcc`? I remember you mentioned `rustup` could be an issue, so I guess you would manually download the Rust corresponding to the version in the [`rust-toolchain` file](https://github.com/rust-lang/rustc_codegen_gcc/blob/master/rust-toolchain). Is this correct?
 * Are you OK with building our [fork of GCC](https://github.com/rust-lang/gcc) and use that? If not, which version of GCC do you have? If we use the upstream GCC, we could add a Cargo feature so that you can get as many features as there are in that GCC version, but there will be a few features missing: especially, `unwinding` is unlikely to get merged upstream soon as we don't support it correctly yet in `rustc_codegen_gcc`. Would you be OK with that?
* Is there more pressure because there are more and more packages for Rust in Debian? If so, what targets do you need? We can probably use that as an argument so that they accept these as new architectures.
* Are you OK with using unstable Rust for now? I don't think alternative codegen backends were stabilized yet.

We should probably make this a MCP or project goal: what do you think, @GuillaumeGomez?

**Comment by glaubitz** [comment]
> I'd say the situation got worse recently since it is now required that the `object` crate supports the architecture for us to be able to compile to an architecture not supported by `rustc_codegen_llvm`. But, I saw you just made a [PR for HPPA](https://github.com/gimli-rs/object/pull/793), so it's a first step in this direction.

Yep. And Alpha and SuperH are already there as well.

> Let me try to find the right people to communicate to and ask about this, but I need to know a few information about what exactly you want:
> 
>     * How do you intend to build `rustc_codegen_gcc`? I remember you mentioned `rustup` could be an issue, so I guess you would manually download the Rust corresponding to the version in the [`rust-toolchain` file](https://github.com/rust-lang/rustc_codegen_gcc/blob/master/rust-toolchain). Is this correct?

I have never used `rustup` and don't plan to use it in the future. I just want to be able to cross-compile a `rustc` host compiler for Alpha, HPPA, M68k and SuperH and create a Debian package from it.

>     * Are you OK with building our [fork of GCC](https://github.com/rust-lang/gcc) and use that? If not, which version of GCC do you have? If we use the upstream GCC, we could add a Cargo feature so that you can get as many features as there are in that GCC version, but there will be a few features missing: especially, `unwinding` is unlikely to get merged upstream soon as we don't support it correctly yet in `rustc_codegen_gcc`. Would you be OK with that?

Local patches are okay, but I would prefer to be able to use Debian's GCC package. We could add the necessary patches there though.

>     * Is there more pressure because there are more and more packages for Rust in Debian? If so, what targets do you need? We can probably use that as an argument so that they accept these as new architectures.
> 
>     * Are you OK with using unstable Rust for now? I don't think alternative codegen backends were stabilized yet.

Yes, of course.

I would argue that getting `rustc` bootstrapped on these remaining four architecture would remove another roadblocker for adding Rust code to more upstream projects.

**Comment by GuillaumeGomez** [comment]
> We should probably make this a MCP or project goal: what do you think, [@GuillaumeGomez](https://github.com/GuillaumeGomez)?

Sounds good to me.

**Comment by glaubitz** [comment]
Adding some more folk who are interested in this topic.

CC @danglin44 @FractalFir @hdeller @sozud

**Comment by thesamesam** [comment]
Whatever works for @glaubitz will surely work for us in Gentoo as well, unless I note otherwise.

Yes, there's more pressure - especially with various Python packages gaining Rust extensions (sometimes non-optional), but also git considering [adoption](https://lore.kernel.org/git/pull.1980.git.git.1752784344.gitgitgadget@gmail.com/) of Rust.

**Comment by antoyo** [comment]
> I just want to be able to cross-compile a `rustc` host compiler for Alpha, HPPA, M68k and SuperH and create a Debian package from it.

Ok, that was not done yet, but @FractalFir works on bootstrapping `rustc` on other architectures for this Google Summer of Code. The first architecture that he will try is Dec Alpha.

Would you be OK if you use a pinned commit of `rustc`? I'm thinking about ways to make sure it's not going to break for you (we currently only run a few tests for cg_gcc in the Rust repo; though, we work on improving that).

> Local patches are okay, but I would prefer to be able to use Debian's GCC package. We could add the necessary patches there though.

Ok, do you want to try first with the patches or using Debian's GCC package? If the latter, I would need the version and will need to add a feature for that GCC version.

**Comment by glaubitz** [comment]
> Ok, that was not done yet, but [@FractalFir](https://github.com/FractalFir) works on bootstrapping `rustc` on other architectures for this Google Summer of Code. The first architecture that he will try is Dec Alpha.

This is awesome! Kudos to @FractalFir!

> Would you be OK if you use a pinned commit of `rustc`? I'm thinking about ways to make sure it's not going to break for you (we currently only run a few tests for cg_gcc in the Rust repo; though, we work on improving that).

Absolutely. I'll take anything I can get here. ;-)

> 
> > Local patches are okay, but I would prefer to be able to use Debian's GCC package. We could add the necessary patches there though.
> 
> Ok, do you want to try first with the patches or using Debian's GCC package? If the latter, I would need the version and will need to add a feature for that GCC version.

Let's try the easy path first. I can add the patches to Debian's GCC package later.

**Comment by antoyo** [comment]
> Let's try the easy path first. I can add the patches to Debian's GCC package later.

I just thought about this: we might require GCC patches to be able to bootstrap correctly on some targets. @FractalFir would probably be able to confirm this.

**Comment by antoyo** [comment]
@glaubitz: I believe opening a [Major Change Proposal (MCP)](https://rust-lang.github.io/rfcs/2904-compiler-major-change-process.html#major-change-proposals) would be the best way forward to make things moving forward for adding the tier-3 support for targets not supported by LLVM.
At the very least, that should give you good discussions with the right people, redirect the discussion to a better place if needed, allow the people to give their concerns, and make the project commit about this.

I suggest you start with this.

**Comment by FractalFir** [comment]
> > Ok, that was not done yet, but [@FractalFir](https://github.com/FractalFir) works on bootstrapping `rustc` on other architectures for this Google Summer of Code. The first architecture that he will try is Dec Alpha.
> 
> This is awesome! Kudos to [@FractalFir](https://github.com/FractalFir)!
> 
> > Would you be OK if you use a pinned commit of `rustc`? I'm thinking about ways to make sure it's not going to break for you (we currently only run a few tests for cg_gcc in the Rust repo; though, we work on improving that).
> 
> Absolutely. I'll take anything I can get here. ;-)
> 
> > > Local patches are okay, but I would prefer to be able to use Debian's GCC package. We could add the necessary patches there though.
> > 
> > 
> > Ok, do you want to try first with the patches or using Debian's GCC package? If the latter, I would need the version and will need to add a feature for that GCC version.
> 
> Let's try the easy path first. I can add the patches to Debian's GCC package later.

Do you have a compleate list of arches you'd like to see supported? Some of the work(eg. making cc-rs recognize those targets) can be easily done for more than one target at a time.

**Comment by glaubitz** [comment]
> Do you have a compleate list of arches you'd like to see supported?

Yes, looking at the [current build status of the `rustc` package in Debian](https://buildd.debian.org/status/package.php?p=rustc&suite=experimental), we're missing the following four architectures:

- alpha
- hppa
- m68k
- sh4

The Rust compiler does actually already support m68k but it's not fully working yet as the LLVM backend still need some work.

> Some of the work(eg. making cc-rs recognize those targets) can be easily done for more than one target at a time.

That would be great and highly appreciated.

Please let me know if there is anything I can help you with!

**Comment by darcagn** [comment]
> Do you have a compleate list of arches you'd like to see supported? Some of the work(eg. making cc-rs recognize those targets) can be easily done for more than one target at a time.

It'd be excellent to get SH4 support! I've been maintaining out-of-tree Rust `std`/`libc` ports to KallistiOS (embedded OS for the SH4-based Sega Dreamcast) using rustc_codegen_gcc and (as mentioned above in 2024) hacks to make `rustc` believe it's work with MIPS. It's been working but difficult for end users to build, sometimes error prone due to ABI behavior mismatches, and I can't upstream the work like this.

As Adrian mentioned above, your work is highly appreciated and let me know if I can help.

**Comment by FractalFir** [comment]
> > Do you have a compleate list of arches you'd like to see supported? Some of the work(eg. making cc-rs recognize those targets) can be easily done for more than one target at a time.
> 
> It'd be excellent to get SH4 support! I've been maintaining out-of-tree Rust `std`/`libc` ports to KallistiOS (embedded OS for the SH4-based Sega Dreamcast) using rustc_codegen_gcc and (as mentioned above in 2024) hacks to make `rustc` believe it's work with MIPS. It's been working but difficult for end users to build, sometimes error prone due to ABI behavior mismatches, and I can't upstream the work like this.
> 
> As Adrian mentioned above, your work is highly appreciated and let me know if I can help.

After some talks to the Rust folk, in order to go forward with any compiler changes, we'd need to create a Major Change Proposal adding a tier 3 target that uses one of those ISAs.

So, if you want support for Dec Alpha or SuperH(or any other ISA), you'd need to create a MCP for those. Here is a[n example of such an MCP.](https://github.com/rust-lang/compiler-team/issues/894)  

This would effectively get us a stamp of aproval from the people in charge, and allow the changes to land in the upstream compiler.

**Comment by FractalFir** [comment]
> sometimes error prone due to ABI behavior mismatches, and I can't upstream the work like this.

Is there an official description of that ABI? That would be needed(or at least helpfull) to add support for it to Rust.

**Comment by FractalFir** [comment]
As for potentially dividing up the work: 

It would be very helpful if the people familiar with the targets made the needed libc changes, and created [the target files.](https://docs.rust-embedded.org/embedonomicon/custom-target.html). That is the area I am least familiar with, and one where your expertise would matter most. (This has to be done after the MCP).

I *think* I can handle most of the compiler-side changes: so far, they don't seem too difficult. 

WRT KallistiOS, could you share your `ct-ng` config? Since it is known to work, I would start with that.

**Comment by thesamesam** [comment]
> After some talks to the Rust folk, in order to go forward with any compiler changes, we'd need to create a Major Change Proposal adding a tier 3 target that uses one of those ISAs.

I fear this is a bit circular. I may be willing to either maintain or co-maintain a target, but I can't really get my feet wet with learning how to integrate it into Rust when I can't get anything landed. I don't want to make a major proposal change when I can't do some baby steps first...

Also, https://doc.rust-lang.org/beta/rustc/target-tier-policy.html#tier-3-target-policy says:
> Tier 3 targets must be able to produce assembly using at least one of rustc's supported backends from any host target. (Having support in a fork of the backend is not sufficient, it must be upstream.)

~That one at least sounds hard to achieve if we can't land PRs like that HPPA one first..~ (OK, this part seems fine maybe, since https://github.com/rust-lang/libc/pull/3542#issuecomment-2033135620 implies proposing it with a GCC backend may work, but it still has the issue I mention above?)

**Comment by thesamesam** [comment]
> I don't want to make a major proposal change when I can't do some baby steps first...

Is it doable to locally build rustc with modified versions of the libc crate and so on (any advice/links/examples for that?)? If so, I can look at that.

**Comment by darcagn** [comment]
> Is there an official description of that ABI? That would be needed(or at least helpfull) to add support for it to Rust.

Yes, see document here: https://dreamcast.wiki/wiki/images/9/92/SH4_C_ABI.pdf

> It would be very helpful if the people familiar with the targets made the needed libc changes

libc-0.2 crate is here in the [`libc-0.2-kos` branch](https://github.com/dreamcast-rs/libc/tree/libc-0.2-kos)
It's a little outdated with upstream libc-0.2, but works.

> and created the target files.

Available here: [sh-elf.json](https://github.com/dreamcast-rs/rust-for-dreamcast/blob/master/misc/sh-elf.json)
But of course, it says `"arch": "mips", "llvm-target": "mips-unknown-linux-gnu",` because no SH target exists and I "fix" that with this [linker wrapper](https://github.com/dreamcast-rs/rust-for-dreamcast/blob/master/misc/wrappers/sh-linker-wrapper/src/main.rs)

> WRT KallistiOS, could you share your `ct-ng` config? Since it is known to work, I would start with that.

I haven't used `ct-ng` to build a KallistiOS toolchain, since we have our own toolchain build scripts.
I have instructions on how to set it up available at https://dreamcast.rs/ using supporting files/patches from the repo here: https://github.com/dreamcast-rs/rust-for-dreamcast

Unfortunately, right now one of the components has a version mismatch and following those instructions will fail to compile rustc_codegen_gcc/sysroot at the final step. I need to update the components (last updated in May). But the instructions to build a GCC cross-compiler should work. I am out of office for a week or so but when I get back I can update those components so the instructions work for rustc_codegen_gcc/sysroot compilation again.

**Comment by FractalFir** [comment]
I think it would be best if those concerns were [voiced to the compiler team here](https://rust-lang.zulipchat.com/#narrow/channel/131828-t-compiler/topic/Support.20for.20GCC-only.20architectures/with/530647398).

This process is a bit of a mess in this case, sadly. 

> Is it doable to locally build rustc with modified versions of the libc crate and so on (any advice/links/examples for that?)? If so, I can look at that.

I have a way of doing that, but it is very brittle and cursed(modifying the code in the cargo cache). You should probably ask about this in the [compiler help channel. ](https://rust-lang.zulipchat.com/#narrow/channel/182449-t-compiler.2Fhelp).

I will look into making this easier with cg_gcc... not sure how I'd go about this, tough.  

For GCC-specific questions, you can also ask in the [rustc_codegen_gcc](https://rust-lang.zulipchat.com/#narrow/channel/386786-rustc-codegen-gcc/topic/Cannot.20compile.20rustc.20on.20Windows.2010/with/528583898) channel. 

The people here are quite helpful, so if you have any questions WRT anything related to porting rust to a new target, that is the right place. Questions are usually answered within an hour or so, sometimes up to a day(when you ask about something very specific that few people know about). 

If it helps, I can outline the list of changes I made to build for DEC-alpha.

**Comment by glaubitz** [comment]
[`hppa-unknown-linux-gnu` was already added in the past to `libc`](https://github.com/rust-lang/libc/pull/3539) but had to be [removed again due to lack of compiler support.](https://github.com/rust-lang/libc/pull/3542).

`m68k-unknown-linux-gnu` is already supported by `libc`.

We're just missing `alpha-unknown-linux-gnu` then.

**Comment by glaubitz** [comment]
> > sometimes error prone due to ABI behavior mismatches, and I can't upstream the work like this.
> 
> Is there an official description of that ABI? That would be needed(or at least helpfull) to add support for it to Rust.

Most SysV ABI specifications can be found [here](https://uclibc.org/specs.html).

**Comment by FractalFir** [comment]
> > > sometimes error prone due to ABI behavior mismatches, and I can't upstream the work like this.
> > 
> > 
> > Is there an official description of that ABI? That would be needed(or at least helpfull) to add support for it to Rust.
> 
> Most SysV ABI specifications can be found [here](https://uclibc.org/specs.html).

Right, I just don't know what ABI KallistiOS uses. That could pose a problem.

**Comment by darcagn** [comment]
> Right, I just don't know what ABI KallistiOS uses. That could pose a problem.

KallistiOS supports using the `m4-single` and `m4-single-only` ABIs (see [SH Options](https://gcc.gnu.org/onlinedocs/gcc/SH-Options.html#index-m4a-single-only)). The specification document is [linked above](https://github.com/rust-lang/rustc_codegen_gcc/issues/49#issuecomment-3115018868).

**Comment by FractalFir** [comment]
Found some info on using a modified `libc`. 

https://rustc-dev-guide.rust-lang.org/building/new-target.html#patching-crates
```
 [patch.crates-io]
libc = {path = "my_libc_path"}
```

You should be able to use this modified libc without Rust compiler changes.

**Comment by glaubitz** [comment]
> > > > sometimes error prone due to ABI behavior mismatches, and I can't upstream the work like this.
> > > 
> > > 
> > > Is there an official description of that ABI? That would be needed(or at least helpfull) to add support for it to Rust.
> > 
> > 
> > Most SysV ABI specifications can be found [here](https://uclibc.org/specs.html).
> 
> Right, I just don't know what ABI KallistiOS uses. That could pose a problem.

I suggest implementing the Linux ABI first and then add other ABIs as necessary.

**Comment by antoyo** [comment]
> I fear this is a bit circular. I may be willing to either maintain or co-maintain a target, but I can't really get my feet wet with learning how to integrate it into Rust when I can't get anything landed. I don't want to make a major proposal change when I can't do some baby steps first...

Ok, let's focus on one target at a time.
I discussed with @FractalFir and we'll focus on building a rust compiler for m68k since the target is already supported in Rust and as such, it should be the easiest.

@glaubitz: How do you intend to build the rust compiler to create the Debian package? Would you use the bootstrap script ([`x`](https://github.com/rust-lang/rust/blob/master/x))? Or do you need/use another process?

**Comment by glaubitz** [comment]
> > I fear this is a bit circular. I may be willing to either maintain or co-maintain a target, but I can't really get my feet wet with learning how to integrate it into Rust when I can't get anything landed. I don't want to make a major proposal change when I can't do some baby steps first...
> 
> Ok, let's focus on one target at a time. I discussed with [@FractalFir](https://github.com/FractalFir) and we'll focus on building a rust compiler for m68k since the target is already supported in Rust and as such, it should be the easiest.

Fine with me. You can also default to 32-bit alignment as this is also the proper alignment [according to the official ABI](https://people.debian.org/~glaubitz/m68k-sysv-abi.pdf). See also [this page on the Debian Wiki](https://wiki.debian.org/M68k/Alignment).

> [@glaubitz](https://github.com/glaubitz): How do you intend to build the rust compiler to create the Debian package? Would you use the bootstrap script ([`x`](https://github.com/rust-lang/rust/blob/master/x))? Or do you need/use another process?

I'm happy with anything you can provide. Of course, it would be perfect if I could just create a host compiler by running `x.py` but I expect this to still take some time.
